src/python/pants/core/goals/deploy.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from abc import ABCMeta
+from dataclasses import dataclass
+from itertools import chain
+
+from pants.core.goals.publish import PublishFieldSet, PublishProcesses, PublishProcessesRequest
+from pants.engine.collection import Collection
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import InteractiveProcess, InteractiveProcessResult
+from pants.engine.rules import Effect, Get, MultiGet, collect_rules, goal_rule, rule
+from pants.engine.target import (
+    FieldSet,
+    NoApplicableTargetsBehavior,
+    TargetRootsToFieldSets,
+    TargetRootsToFieldSetsRequest,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import union
+from pants.option.option_types import ArgsListOption
+
+logger = logging.getLogger(__name__)
+
+
+@union
+@dataclass(frozen=True)
+class DeployFieldSet(FieldSet, metaclass=ABCMeta):
+    """The FieldSet type for the `deploy` goal."""
+
+
+@dataclass(frozen=True)
+class DeployProcess:
+    """Individual process to run in the given order to perform the deployment."""
+
```

Comment:
As mentioned on the document, individual implementations should likely be able to declare publish and deploy dependencies here, rather than necessarily always using the transitive dependencies.